### DISCUSSION TITLE
请问在一个会话中依赖注入的session是同一个吗？session commit后需不需要手动close()

### DISCUSSION THREAD

**Discussion by wiketool** [he0119/nonebot-plugin-datastore#157] [Q&A] [answered]
请问在一个会话中依赖注入的session是同一个吗？session commit后需不需要手动close()

**Comment by he0119** [comment, marked as answer]
是同一个（会缓存），并且不需要手动关闭，会在 matcher 结束后自动关闭。

<https://nonebot.dev/docs/advanced/dependency#%E4%BE%9D%E8%B5%96%E7%BC%93%E5%AD%98>
<https://nonebot.dev/docs/advanced/dependency#%E7%94%9F%E6%88%90%E5%99%A8%E4%BD%9C%E4%B8%BA%E4%BE%9D%E8%B5%96>

**Reply by wiketool** [reply]
补充一个提问，create_session()在会话外创建的session需要手动关闭吗？代码如下，麻烦了
```python
async with create_session() as session:
    session.add(record)
    await session.commit()
```

**Reply by he0119** [reply]
你用了 with 的话，就会自动关闭的。没有的话，就需要手动关闭。

**Reply by he0119** [reply]
具体可以参考文档 <https://docs.sqlalchemy.org/en/20/orm/session_basics.html#opening-and-closing-a-session>。

**Comment by wiketool** [comment]
好的谢谢


he0119 ***@***.***>于2023年7月23日 周日22:28写道：
